### PR TITLE
Add check when trying to reroute a shard to a non-data discovery node

### PR DESCRIPTION
**Pull request by Autophagy**
This was causing a NPE as the toDiscoNode was not in the list of routingNodes.

### CODE REVIEW

**Comment by Autophagy** [comment]
Also, does this need to be backported at all to any other cherry branches? Do I also have to make a PR against crate/crate to change the upstream? Been a while since I did anything with ES :sweat_smile:

**Comment by mfussenegger** [comment]
> Also, does this need to be backported at all to any other cherry branches?

Yep, actually `v6.1.3_cherry` and `v5.6.3_cherry`  are currently in use (`master` and `2.3`)

> Do I also have to make a PR against crate/crate to change the upstream

Yep, please. But I'd do it after the ES change has been merged

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/MoveAllocationCommand.java [hunk 1]
```diff
@@ -122,6 +122,9 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
             }
 
             RoutingNode toRoutingNode = allocation.routingNodes().node(toDiscoNode.getId());
+            if (toRoutingNode == null && !toDiscoNode.isDataNode()) {
```
Comment (mfussenegger) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/MoveAllocationCommand.java]: is the `allocation.routingNodes().node(fromDiscoNode.getId())` call safe that is happening further above in the `for`  statement?

Comment (Autophagy) [reply core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/MoveAllocationCommand.java]: Good spot, it's not. Will add.

**Comment by Autophagy** [comment]
@mfussenegger Do you have a test I could look at for how I would test this? Been looking through the existing tests but there's quite a lot and I'm having trouble seeing where theres any existing move allocation command tests

**Comment by mfussenegger** [comment]
Not really, but you could look into creating a unit test for the `MoveAllocationCommand.execute` directly.

In our `SQLExecutor` is an example how to build a `AllocationService` which should allow you to build a dummy `RoutingAllocation`

**Comment by Autophagy** [comment]
Awesome, thanks a bunch :slightly_smiling_face:

**Pull request merged**